### PR TITLE
Database

### PR DESCRIPTION
**Pull request by hqzhu7**
Including everything to use Hibernate(regions.classes, configuration file, help classes, also Algorithm and Regiongrowing)

### CODE REVIEW

**Comment by hqzhu7** [comment]
the main code is under "Updated master"

File: src/main/java/algorithms/RegionGrowing.java [hunk 1]
```diff
@@ -1,21 +1,127 @@
...
+    WKTWriter writer = new WKTWriter();
+    //Compactness calculator
+    Schwartzberg szg = new Schwartzberg();
+    
+    //every time add a precinct to the district, newDistrictBoundary grows
+    HashSet<Geometry> newDistrictBoundary = new HashSet<>();
+    //every time add a precinct to the district, also add it's adjPrecncts(ID) to this list as candidates
+    HashSet<Precinct> candidatePrecincts = new HashSet<>();
+
+    //get a seed to start
```
Comment (JaysonDeMarchi) [inline src/main/java/algorithms/RegionGrowing.java]: A lot of this logic is going to have to get moved around.
Start algorithm should only handle initialization and a call to the first iteration of `run()`.
There's logic in here that needs to be moved to the metrics. Well work on this tomorrow.

File: src/main/java/algorithms/RegionGrowing.java [hunk 2]
```diff
@@ -1,21 +1,127 @@
...
+   * This is used to get precincts randomly to start region growing according to
+   * existing districts.
+   *
+   * @param districts
+   */
+  public void setSeedsRandomly(Collection<District> districts) {
+    this.seeds = new ArrayList<>();
+    int i=0;
+    for (District d : districts) {
+      this.seeds.add(((List<Precinct>) (List) d.getPrecincts()).get(i+=100));
```
Comment (JaysonDeMarchi) [inline src/main/java/algorithms/RegionGrowing.java]: This isn't random this is very evenly spaced on the arrayList.
Let's actually randomize this. I have a branch that impliments something similar we can use.

File: src/main/java/utils/Validator.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
+/*
```
Comment (JaysonDeMarchi) [inline src/main/java/utils/Validator.java]: Let's remove this commented out instructions for setting up a header.

File: src/main/java/utils/Validator.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * This function is used to tests if a string is a JSON string or not .
+ *
+ * @Param test The String to test.
+ * @Return boolean This returns test is validate.
+ * @author Hengqi Zhu
+ */
+public class Validator {
+
+  public static boolean isJSONValid(String test) {
```
Comment (JaysonDeMarchi) [inline src/main/java/utils/Validator.java]: I'm not sure where we would be needing this.

File: src/main/java/utils/HibernateManager.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  /**
+   *
+   * @return instance of HibernateManager
+   * @throws HibernateException
+   */
+  public static synchronized HibernateManager getInstance() throws Exception {
+    if (instance == null) {
+      instance = new HibernateManager();
+    }
+    return instance;
```
Comment (JaysonDeMarchi) [inline src/main/java/utils/HibernateManager.java]: Use `this`

File: src/main/java/servlets/StartAlgorithm.java [hunk 6]
```diff
@@ -33,6 +36,11 @@
   protected void processRequest(HttpServletRequest request, HttpServletResponse response)
           throws ServletException, IOException {
     BufferedReader br = request.getReader();
+    
+    
+    Map<Metric, Float> criteria = new HashMap<>();
+    RegionGrowing rg = new RegionGrowing(UT, criteria);
```
Comment (JaysonDeMarchi) [inline src/main/java/servlets/StartAlgorithm.java]: Guessing this is your test?
We don't want to merge test code.

File: src/main/java/regions/Precinct.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+  public String getDistrictId() {
+    return this.districtId;
+  }
+
+  public void setDistrictId(String districtId) {
+    this.districtId = districtId;
+  }
+
+  @Column(name = "ADJPRECINCTS")
+  public String getAdjPrecincts() {
```
Comment (JaysonDeMarchi) [inline src/main/java/regions/Precinct.java]: This should be a `Collection<Precincts>`.
We may be receiving the data as a String but we want to parse the data as useful objects, or else we will have to convert it every time we use this attribute.

**Pull request merged**